python/runtime/cudaq/algorithms/py_resource_count.cpp

```diff
@@ -0,0 +1,68 @@
+/*******************************************************************************
+ * Copyright (c) 2022 - 2025 NVIDIA Corporation & Affiliates.                  *
+ * All rights reserved.                                                        *
+ *                                                                             *
+ * This source code and the accompanying materials are made available under    *
+ * the terms of the Apache License 2.0 which accompanies this distribution.    *
+ ******************************************************************************/
+
+#include "runtime/cudaq/platform/py_alt_launch_kernel.h"
+#include "utils/LinkedLibraryHolder.h"
+#include "utils/OpaqueArguments.h"
+#include "mlir/Bindings/Python/PybindAdaptors.h"
+#include "mlir/CAPI/IR.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+
+#include <fmt/core.h>
+#include <pybind11/functional.h>
+#include <pybind11/stl.h>
+
+namespace py = pybind11;
+
+namespace cudaq {
+void pyAltLaunchKernel(const std::string &, MlirModule, OpaqueArguments &,
+                       const std::vector<std::string> &);
+
+void bindCountResources(py::module &mod) {
+  mod.def(
+      "count_resources",
+      [&](std::function<bool()> choice, py::object kernel, py::args args) {
+        if (py::hasattr(kernel, "compile"))
+          kernel.attr("compile")();
+        auto &platform = cudaq::get_platform();
+        auto kernelName = kernel.attr("name").cast<std::string>();
+        auto kernelMod = kernel.attr("module").cast<MlirModule>();
+        args = simplifiedValidateInputArguments(args);
+        auto *argData = toOpaqueArgs(args, kernelMod, kernelName);
+
+        auto ctx = std::make_unique<ExecutionContext>("resourcecount", 1);
+        ctx->kernelName = kernelName;
+        ctx->choice = choice;
```

Comment:
I am wondering - can the choice function be directly communicated to the resource counting simulator, to avoid extra fields on execution context?

Reply:
I think that would require having direct access to the simulator in the `count_resources` runtime functions for both C++ and python, which should be doable but may be a bit tricky.

Reply:
Looking into it, I'm not sure how doable this is, because there's currently a linking issue that prevents directly referencing the ResourceCounter simulator in the C++ `count_resources` function, which we would have to solve or work around.